### ISSUE TITLE
Update code for compatibility with PHP8

### ISSUE THREAD

**Issue by JimBacon** [Indicia-Team/client_helpers#118]
Relates to BiologicalRecordsCentre/iRecord#1047

**Comment by JimBacon** [comment]
Running PHP Compatibility checks for PHP >= 7.3 against the `dev` branch with
`phpcs -p --standard=PHPCompatibility --runtime-set testVersion 7.3- --report=summary client_helpers`
returns the following summary results.
```
--------------------------------------------------------------------------------
FILE                                                            ERRORS  WARNINGS
--------------------------------------------------------------------------------
client_helpers/data_entry_helper.php                                0       2
client_helpers/lang.php                                             0       1
client_helpers/secure_msg.php                                       33      0
client_helpers/prebuilt_forms/mnhnl_dynamic_2.php                   2       0
client_helpers/prebuilt_forms/ukbms_section_plot.php                0       1
client_helpers/prebuilt_forms/wwt_colour_marked_report.php          1       0
--------------------------------------------------------------------------------
A TOTAL OF 36 ERRORS AND 4 WARNINGS WERE FOUND IN 6 FILES
--------------------------------------------------------------------------------
```

**Comment by JimBacon** [comment]
The errors details are given by
`phpcs -pn --standard=PHPCompatibility --runtime-set testVersion 7.3- --report-width=100 client_helpers`
which yields
```
FILE: client_helpers/prebuilt_forms/wwt_colour_marked_report.php
----------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------------------------------------
 1833 | ERROR | The left-associativity of the ternary operator has been deprecated in PHP 7.4 and
      |       | removed in PHP 8.0. Multiple consecutive ternaries detected. Use parenthesis to
      |       | clarify the order in which the operations should be executed
----------------------------------------------------------------------------------------------------


FILE: client_helpers/prebuilt_forms/mnhnl_dynamic_2.php
----------------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------------------------------------
 517 | ERROR | Using 'break' outside of a loop or switch structure is invalid and will throw a
     |       | fatal error since PHP 7.0
 519 | ERROR | Using 'continue' outside of a loop or switch structure is invalid and will throw a
     |       | fatal error since PHP 7.0
----------------------------------------------------------------------------------------------------


FILE: client_helpers/secure_msg.php
----------------------------------------------------------------------------------------------------
FOUND 33 ERRORS AFFECTING 12 LINES
----------------------------------------------------------------------------------------------------
 147 | ERROR | Function mcrypt_module_open() is deprecated since PHP 7.1 and removed since PHP 7.2;
     |       | Use OpenSSL instead
 147 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | openssl (preferred) or pecl/mcrypt once available instead
 147 | ERROR | The constant "MCRYPT_RIJNDAEL_256" is deprecated since PHP 7.1 and removed since PHP
     |       | 7.2
 147 | ERROR | The constant "MCRYPT_MODE_ECB" is deprecated since PHP 7.1 and removed since PHP 7.2
 148 | ERROR | Function mcrypt_create_iv() is deprecated since PHP 7.1 and removed since PHP 7.2;
     |       | Use random_bytes() or OpenSSL instead
 148 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | openssl (preferred) or pecl/mcrypt once available instead
 148 | ERROR | Function mcrypt_enc_get_iv_size() is deprecated since PHP 7.1 and removed since PHP
     |       | 7.2; Use OpenSSL instead
 148 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | openssl (preferred) or pecl/mcrypt once available instead
 148 | ERROR | The constant "MCRYPT_RAND" is deprecated since PHP 7.1 and removed since PHP 7.2
 149 | ERROR | Function mcrypt_generic_init() is deprecated since PHP 7.1 and removed since PHP
     |       | 7.2; Use OpenSSL instead
 149 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | openssl (preferred) or pecl/mcrypt once available instead
 150 | ERROR | Function mcrypt_generic() is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | OpenSSL instead
 150 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | openssl (preferred) or pecl/mcrypt once available instead
 151 | ERROR | Function mcrypt_generic_deinit() is deprecated since PHP 7.1 and removed since PHP
     |       | 7.2; Use OpenSSL instead
 151 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | openssl (preferred) or pecl/mcrypt once available instead
 152 | ERROR | Function mcrypt_module_close() is deprecated since PHP 7.1 and removed since PHP
     |       | 7.2; Use OpenSSL instead
 152 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | openssl (preferred) or pecl/mcrypt once available instead
 170 | ERROR | Function mcrypt_module_open() is deprecated since PHP 7.1 and removed since PHP 7.2;
     |       | Use OpenSSL instead
 170 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | openssl (preferred) or pecl/mcrypt once available instead
 170 | ERROR | The constant "MCRYPT_RIJNDAEL_256" is deprecated since PHP 7.1 and removed since PHP
     |       | 7.2
 170 | ERROR | The constant "MCRYPT_MODE_ECB" is deprecated since PHP 7.1 and removed since PHP 7.2
 171 | ERROR | Function mcrypt_create_iv() is deprecated since PHP 7.1 and removed since PHP 7.2;
     |       | Use random_bytes() or OpenSSL instead
 171 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | openssl (preferred) or pecl/mcrypt once available instead
 171 | ERROR | Function mcrypt_enc_get_iv_size() is deprecated since PHP 7.1 and removed since PHP
     |       | 7.2; Use OpenSSL instead
 171 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | openssl (preferred) or pecl/mcrypt once available instead
 171 | ERROR | The constant "MCRYPT_RAND" is deprecated since PHP 7.1 and removed since PHP 7.2
 172 | ERROR | Function mcrypt_generic_init() is deprecated since PHP 7.1 and removed since PHP
     |       | 7.2; Use OpenSSL instead
 172 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | openssl (preferred) or pecl/mcrypt once available instead
 173 | ERROR | Function mdecrypt_generic() is deprecated since PHP 7.1 and removed since PHP 7.2;
     |       | Use OpenSSL instead
 174 | ERROR | Function mcrypt_generic_deinit() is deprecated since PHP 7.1 and removed since PHP
     |       | 7.2; Use OpenSSL instead
 174 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | openssl (preferred) or pecl/mcrypt once available instead
 175 | ERROR | Function mcrypt_module_close() is deprecated since PHP 7.1 and removed since PHP
     |       | 7.2; Use OpenSSL instead
 175 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | openssl (preferred) or pecl/mcrypt once available instead
----------------------------------------------------------------------------------------------------
```

**Comment by johnvanbreda** [comment]
Before making changes in secure_msg.php, I would be tempted to look at whether the user_helper class (which uses it) is actually required in any active projects, as this may be additional unnecessary complexity.

**Comment by JimBacon** [comment]
I'm not sure I have any way of knowing where there are active projects or what they are using. There is a polyfill, https://github.com/phpseclib/mcrypt_compat, which I have working.

**Comment by johnvanbreda** [comment]
The string "user_helper" only occurs in 1 place which is the class definition. I've searched all the code on the warehouse, which includes client_helpers and it's the same for my Drupal 8 install. The file has never had any commits other than generic code updates since it was added to GitHub. It's also not documented on Read the Docs. 

So, I think this file represented an early attempt to allow warehouse-managed login on PHP only Indicia setups (i.e. outside Drupal). There were only ever a couple of those, like RISC, but I don't recall any using the warehouse to manage logins.

I agree though we should try to be a little more certain about it not being used before dropping it, but on the flip-side, any reduction in complexity has to be a good thing.

**Comment by JimBacon** [comment]
The back-end for the user_helper is the `/services/site_user/authenticate_user` end point which also uses `secure_msg.php`. I wondered if this was being used by any apps, for example, as a means of authentication.

I'm very much in favour of reducing complexity if we can!

**Comment by johnvanbreda** [comment]
Good spot @JimBacon. Again though, I can't find anything using that end-point other than the user_helper class and it's undocumented. I suspect it was an early effort to supply user authentication on the warehouse which got superseded by the Drupal integration. It doesn't seem to have been updated since 2012 other than generic code changes.

Whilst I'm 90% happy it's unused, it would be nice to be a little more certain before removing it.

**Comment by BirenRathod** [comment]
@johnvanbreda & @JimBacon More on this. The PHP 8.0.13 was running fine on dev warehouse though I haven't checked import/export features on it but shouldn't be a problem.

So, after 8.1 release, I just installed 8.1.3 last week and tried running warehouse using that version and I found below error.
PHP Fatal error:  During inheritance of ArrayAccess: Uncaught ErrorException: Return type of Database_Result::offsetExists($offset) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in D:\websites\warehouse\application\libraries\drivers\Database.php:585
